src/bvals/bvals_swarm.cpp

```diff
@@ -0,0 +1,166 @@
+  std::shared_ptr<MeshBlock> pmb = GetBlockPointer();
+  int &mylevel = pmb->loc.level;
+  // Fence to make sure buffers are loaded before sending
+  pmb->exec_space.fence();
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+    if (nb.snb.rank != Globals::my_rank) {
+#ifdef MPI_PARALLEL
+      PARTHENON_REQUIRE(bd_var_.req_send[nb.bufid] == MPI_REQUEST_NULL,
+                        "Trying to create a new send before previous send completes!");
+      MPI_Isend(bd_var_.send[n].data(), send_size[n], MPI_PARTHENON_REAL, nb.snb.rank,
+                send_tag[n], MPI_COMM_WORLD, &(bd_var_.req_send[nb.bufid]));
+#endif // MPI_PARALLEL
+    } else {
+      MeshBlock &target_block = *pmy_mesh_->FindMeshBlock(nb.snb.gid);
+      std::shared_ptr<BoundarySwarm> ptarget_bswarm =
+          target_block.pbswarm->bswarms[bswarm_index];
+      if (send_size[nb.bufid] > 0) {
+        // Ensure target buffer is large enough
+        if (bd_var_.send[nb.bufid].extent(0) >
+            ptarget_bswarm->bd_var_.recv[nb.targetid].extent(0)) {
+          ptarget_bswarm->bd_var_.recv[nb.targetid] =
+              ParArray1D<Real>("Buffer", (bd_var_.send[nb.bufid].extent(0)));
+        }
+
+        target_block.deep_copy(ptarget_bswarm->bd_var_.recv[nb.targetid],
+                               bd_var_.send[nb.bufid]);
+        ptarget_bswarm->recv_size[nb.targetid] = send_size[nb.bufid];
+        ptarget_bswarm->bd_var_.flag[nb.targetid] = BoundaryStatus::arrived;
+      } else {
+        ptarget_bswarm->recv_size[nb.targetid] = 0;
+        ptarget_bswarm->bd_var_.flag[nb.targetid] = BoundaryStatus::completed;
+      }
+    }
+  }
+}
+
+void BoundarySwarm::Receive(BoundaryCommSubset phase) {
+#ifdef MPI_PARALLEL
+  MPI_Barrier(MPI_COMM_WORLD);
+  std::shared_ptr<MeshBlock> pmb = GetBlockPointer();
+  int &mylevel = pmb->loc.level;
+  for (int n = 0; n < pmb->pbval->nneighbor; n++) {
+    NeighborBlock &nb = pmb->pbval->neighbor[n];
+    if (nb.snb.rank != Globals::my_rank) {
+      pmb->exec_space.fence();
+      // Check to see if we got a message
+      int test;
+      MPI_Status status;
+
+      if (bd_var_.flag[nb.bufid] != BoundaryStatus::completed) {
+        MPI_Iprobe(MPI_ANY_SOURCE, recv_tag[nb.bufid], MPI_COMM_WORLD, &test, &status);
+        if (!static_cast<bool>(test)) {
+          bd_var_.flag[nb.bufid] = BoundaryStatus::waiting;
+        } else {
+          bd_var_.flag[nb.bufid] = BoundaryStatus::arrived;
+
+          // If message is available, receive it
+          int nbytes = 0;
+          MPI_Get_count(&status, MPI_CHAR, &nbytes);
+          if (nbytes / sizeof(Real) > bd_var_.recv[n].extent(0)) {
+            bd_var_.recv[n] = ParArray1D<Real>("Buffer", nbytes / sizeof(Real));
+          }
+          MPI_Recv(bd_var_.recv[n].data(), nbytes, MPI_CHAR, nb.snb.rank,
+                   recv_tag[nb.bufid], MPI_COMM_WORLD, &status);
+          recv_size[n] = nbytes / sizeof(Real);
```

Comment:
Technically you should be using `MPI_BYTE` if you want to type erase the buffers, but why not just use `MPI_PARTHENON_REAL` here?

Reply:
Yes right I can just use `MPI_PARTHENON_REAL`, I wasn't paying attention to how `MPI_Get_count()` works. Done.